Dev v0.1.0 Land Surface Temperature changes

raydittmeier opened the pull request:
This includes a significant reorganization of the LST main processing code, switching some phases from one language to another (e.g.: C to Python), and avoiding processing grid points that are not needed/used.

ronaldddilley commented on not-validated-prototype_lst/scripts/lst_run_modtran.py:
```diff
@@ -0,0 +1,268 @@
+    return args
+
+
+class Tape6rocessingError(Exception):
```
Typo in exception name

ronaldddilley commented on not-validated-prototype_lst/scripts/lst_run_modtran.py:
```diff
@@ -0,0 +1,268 @@
+                break
+
+    if value is None:
+        raise Tape6rocessingError('Unable to find [{}] in tape6 file'
```
typo in exception name

ronaldddilley commented on not-validated-prototype_lst/scripts/lst_run_modtran.py:
```diff
@@ -0,0 +1,268 @@
+
+
+def process_point_dir((point_path, modtran_data_path)):
+    """Swim in the pool using the path
```
Again, this same comment was copied :(

ronaldddilley commented on not-validated-prototype_lst/scripts/lst_utilities.py:
```diff
@@ -14,13 +14,14 @@
 
 
-import metadata_api
+#import metadata_api
```
remove this commented import

ronaldddilley commented on not-validated-prototype_lst/src/calculate_atmospheric_parameters.c:
```diff
@@ -0,0 +1,2333 @@
+          e0 = easting of starting point
+          e2 = easting of stopping point
+******************************************************************************/
+double distance_in_utm
```
I think we should keep this distance implementation somewhere as an example of how to do it in UTM,  But we should probably be using the haversine_distance, to be consistent with the python code, and to also allow us to use other projections.  Maybe mixing Haversine in the python and UTM in the C is causing some of the differences.

ronaldddilley commented on not-validated-prototype_lst/src/calculate_atmospheric_parameters.c:
```diff
@@ -0,0 +1,2333 @@
+
+NOTE: The indexes of the grid points are assumed to be populated.
+*****************************************************************************/
+void determine_grid_point_distances
```
This code changes to use the haversine distance.

ronaldddilley commented on not-validated-prototype_lst/src/calculate_atmospheric_parameters.c:
```diff
@@ -0,0 +1,2333 @@
+    -----  -------------------------------------------------------------------
+    index  The index of the center point
+*****************************************************************************/
+int determine_center_grid_point
```
maybe haversine modifications filter through here as well

ronaldddilley commented on not-validated-prototype_lst/src/calculate_atmospheric_parameters.c:
```diff
@@ -0,0 +1,2333 @@
+    -----  -------------------------------------------------------------------
+    index  The index of the center point
+*****************************************************************************/
+int determine_first_center_grid_point
```
maybe haversine modifications filter through here as well

ronaldddilley commented on not-validated-prototype_lst/src/calculate_atmospheric_parameters.c:
```diff
@@ -0,0 +1,2333 @@
+                easting = input->meta.ul_map_corner.x
+                    + (sample * input->x_pixel_size);
+                northing = input->meta.ul_map_corner.y
+                    - (line * input->y_pixel_size);
```
Haversine changes start here probably, need to calculate lat and lon to pass through to haversine code.

raydittmeier commented:
I sent in changes intended to address all of Ron's comments.  We had a
discussion about the ASTER GED login status, and I attempted to update the
README to reflect the status.  The main change was a switch of a UTM-based
function to Haversine to compute the distance between a grid point and a
pixel.  This was tested to show only minor differences compared to previous
results when the distance between a pixel and 2 grid points is close.  (The
differences were in a somewhat sparse "tic-tac-toe" pattern aligned with
the grid points).

On Tue, Dec 13, 2016 at 11:38 AM, Ronald D. Dilley Jr. <
notifications@github.com> wrote:

> *@ronaldddilley* commented on this pull request.
>
> Done with first review
> ------------------------------
>
> In README.md
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > @@ -71,15 +75,15 @@ See `land_surface_temperature.py --help` for command line details.
>  ### Data Processing Requirements
>  This version of the Land Surface Temperature application requires the input products to be in the ESPA internal file format.
>
> -The following input data are required to generate the spectral indicies products:
> +The following input data are required to generate the Land Surface Temperature product:
>  * Top of Atmosphere Reflectance (TOA)
>    - TOA products can be generated using the [LEDAPS](https://github.com/USGS-EROS/espa-surface-reflectance) or [L8_SR](https://github.com/USGS-EROS/espa-surface-reflectance) software found in our [espa-surface-reflectance](https://github.com/USGS-EROS/espa-surface-reflectance) project.  Or through our ondemand processing system [ESPA](https://espa.cr.usgs.gov), be sure to select the ENVI output format.
>  * Elevation
>    - Elevation data is expected to be in the same projection and image dimensions as the TOA products and in the ENVI file format.  The filename is also expected be `<sceneid>_dem.img`: where <b>sceneid</b> is the Landsat Scene ID of the data being processed.
>  * ASTER GED
>    - ASTER GED data can be [found here](https://lpdaac.usgs.gov/data_access/data_pool).  However it will automatically be retrieved for you as needed and cleaned up.
>
> We need to change the ASTER statement. Because for external users, an
> alternate retrieval mechanism is needed. The external access to the data is
> now behind a username and password, where as internal users we have direct
> access to the data.
> ------------------------------
>
> In README.md
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >  * Top of Atmosphere Reflectance (TOA)
>    - TOA products can be generated using the [LEDAPS](https://github.com/USGS-EROS/espa-surface-reflectance) or [L8_SR](https://github.com/USGS-EROS/espa-surface-reflectance) software found in our [espa-surface-reflectance](https://github.com/USGS-EROS/espa-surface-reflectance) project.  Or through our ondemand processing system [ESPA](https://espa.cr.usgs.gov), be sure to select the ENVI output format.
>  * Elevation
>    - Elevation data is expected to be in the same projection and image dimensions as the TOA products and in the ENVI file format.  The filename is also expected be `<sceneid>_dem.img`: where <b>sceneid</b> is the Landsat Scene ID of the data being processed.
>  * ASTER GED
>    - ASTER GED data can be [found here](https://lpdaac.usgs.gov/data_access/data_pool).  However it will automatically be retrieved for you as needed and cleaned up.
>  * North American Regional Reanalysis (NARR)
> -  - NARR data, it would be best to utilize the `lst_aux_data` software provided in this project to download and build your own archive for the dates you require.
> +  - For NARR data, it would be best to utilize the `lst_aux_data` software provided in this project to download and build your own archive for the dates you require.
>
> We need to change this from lst_aux_data to lst_auxiliary_data.
> ------------------------------
>
> In not-validated-prototype_lst/LST-algorithm-description.txt
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >
> -    The DEM is utilized to adjust for the correct height during the
> -    interpolation step.
> +    The ASTER emissivity and NDVI data, as well as the Landsat TOA data
> +        from multiple bands, are used to create an emissivity band matching
>
> Should probably say something like "used to simulate a Landsat emissivity
> band". I like the word simulate, because that is exactly what it is doing.
> ------------------------------
>
> In not-validated-prototype_lst/LST-algorithm-description.txt
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >
>  ==============================================================================
>  Algorithm Description - Detailed:
>
> -TODO TODO TODO - See DSWE for an example.
> +        1) Read Landsat top of atmosphere reflectance green, red, NIR, SWIR1,
> +           and brightness temperature bands for the scene.  Generate a
> +           Landsat NDVI layer from the red and NIR bands.  Generate NDSI and
> +           snow layers from the green and SWIR bands.  Get projection
> +           information from the brightness temperature band.
> +
> +        2) Retrieve ASTER emissivity and NDVI data, and warp them to match
> +           the Landsat scene.
> +
> +        3) Calculate the Landsat emissivity layer using the ASTER emissivity
> +           values, ASTER NDVI values, Landsat NDVI band, NDSI band, and snow
> +           locations.  Write the Landsat emissivity layer to a file.
>
> Should we use the word simulate here as well?
> ------------------------------
>
> In not-validated-prototype_lst/scripts/build_lst_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > @@ -211,22 +214,24 @@ def generate_data(self):
>          # Downwelling sky irradiance
>          self.logger.info('Loading intermediate downwelled band data [{0}]'
>                           .format(self.downwelled_name))
> -        ds = gdal.Open(self.downwelled_name)
> -        downwelled_data = ds.GetRasterBand(1).ReadAsArray(0, 0, x_dim, y_dim)
> +        dataset = gdal.Open(self.downwelled_name)
> +        downwelled_data = dataset.GetRasterBand(1).ReadAsArray(0, 0, x_dim, \
> +            y_dim)
>
> I would refrain from using backslash as much as possible. Mostly because I
> don't like it, but also because it is almost never needed.
>
> The parameters are wrapped with parens, so the backslash isn't needed.
> pylint/pychecker probably complained about indenting of some sort. But all
> that needed to happen to was to indent the y_dim to line up with the "(0,"
> first zero of the parameters. That should remove the pylint/pychecker
> complaint.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/build_lst_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >
>          # Landsat emissivity estimated from ASTER GED data
>          self.logger.info('Loading intermediate emissivity band data [{0}]'
>                           .format(self.emissivity_name))
> -        ds = gdal.Open(self.emissivity_name)
> -        emissivity_data = ds.GetRasterBand(1).ReadAsArray(0, 0, x_dim, y_dim)
> +        dataset = gdal.Open(self.emissivity_name)
> +        emissivity_data = dataset.GetRasterBand(1).ReadAsArray(0, 0, x_dim, \
> +            y_dim)
>
> Same comment as above.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/build_lst_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > @@ -422,7 +426,7 @@ def generate_data(self):
>
>      # Report the version and exit
>      if args.version:
> -        print(util.Version.version_text())
> +        print util.Version.version_text()
>
> This entire if statement isn't needed if we implement the version argument
> correctly for ArgumentParser.
> See cli.py in espa-processing for an example of how to specify the version
> argument.
>
> Just FYI, pylint complains about the parens with print statements, but
> when we migrate to python 3 someday in the future, the print statements
> require parens. I'm assuming I added them for that reason.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >      args = parser.parse_args()
>
>      # Report the version and exit
>      if args.version:
> -        print(util.Version.version_text())
> +        print util.Version.version_text()
>
> Same here as a previous comment with version arguments.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_build_modtran_input.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    parser.add_argument('--version',
> +                        action='store_true', dest='version',
> +                        required=False, default=False,
> +                        help='Reports the version of the software')
> +
> +    args = parser.parse_args()
> +
> +    # Command line arguments are required so print the help if none were
> +    # provided
> +    if len(sys.argv) == 1:
> +        parser.print_help()
> +        sys.exit(1)  # EXIT FAILURE
> +
> +    if args.version:
> +        print util.Version.version_text()
>
> Same version comment as previously mentioned.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_determine_grid_points.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    parser.add_argument('--version',
> +                        action='store_true', dest='version',
> +                        required=False, default=False,
> +                        help='Reports the version of the software')
> +
> +    args = parser.parse_args()
> +
> +    # Command line arguments are required so print the help if none were
> +    # provided
> +    if len(sys.argv) == 1:
> +        parser.print_help()
> +        sys.exit(1)  # EXIT FAILURE
> +
> +    if args.version:
> +        print util.Version.version_text()
>
> Same version comment as before
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_determine_grid_points.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    (lr_lon, lr_lat, dummy) = gdal_objs.data_to_ll.TransformPoint(east_x,
> +                                                                  south_y)
> +    (ll_lon, ll_lat, dummy) = gdal_objs.data_to_ll.TransformPoint(west_x,
> +                                                                  south_y)
> +
> +    north_lat = max(ul_lat, ur_lat)
> +    south_lat = min(lr_lat, ll_lat)
> +    east_lon = max(ur_lon, lr_lon)
> +    west_lon = min(ul_lon, ll_lon)
> +
> +    return DataBoundInfo(north_lat=north_lat, south_lat=south_lat,
> +                         east_lon=east_lon, west_lon=west_lon)
> +
> +
> +def check_within_bounds(data_bounds, lon, lat):
> +    """
>
> I missed adding the short description, please add one.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    parser.add_argument('--version',
> +                        action='store_true', dest='version',
> +                        required=False, default=False,
> +                        help='Reports the version of the software')
> +
> +    # Parse the command line parameters
> +    args = parser.parse_args()
> +
> +    # Command line arguments are required so print the help if none were
> +    # provided
> +    if len(sys.argv) == 1:
> +        parser.print_help()
> +        sys.exit(1)  # EXIT FAILURE
> +
> +    if args.version:
> +        print util.Version.version_text()
>
> Same version comment as before.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    if args.aux_path == '':
> +        raise Exception('No LST Auxiliary data directory provided.')
> +
> +    return args
> +
> +
> +AuxFilenameSet = namedtuple('AuxFilenameSet',
> +                            ('parameter', 'hdr', 'grb', 'output_dir'))
> +
> +
> +def build_aux_filenames(aux_path, parm, date, date_type):
> +    """Builds a filename set based on date and time
> +
> +    Args:
> +        TODO TODO TODO
>
> Please add argument descriptions
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    path = AUX_PATH_TEMPLATE.format(date.year, date.month, date.day)
> +
> +    hdr_path = os.path.join(aux_path, path, filename)
> +
> +    return AuxFilenameSet(parameter=parm,
> +                          hdr=hdr_path,
> +                          grb=hdr_path.replace('.hdr', '.grb'),
> +                          output_dir='{0}_{1}'.format(parm, date_type))
> +
> +
> +def aux_filenames(aux_path, parms, t0_date, t1_date):
> +    """Builds t0 and t1 filename sets for each parameter
> +
> +    Args:
> +        TODO TODO TODO
>
> Please add argument descriptions.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    Returns:
> +        list(<AuxFilenameSet>): A list of the filename sets
> +    """
> +
> +    for parm in parms:
> +        yield build_aux_filenames(aux_path=aux_path, parm=parm,
> +                                  date=t0_date, date_type='t0')
> +        yield build_aux_filenames(aux_path=aux_path, parm=parm,
> +                                  date=t1_date, date_type='t1')
> +
> +
> +def extract_from_grib(aux_set):
> +    """Extracts the information from the grib file into a directory
> +
> +    Args:
> +        TODO TODO TODO
>
> same
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +            output = ''
> +            try:
> +                output = util.System.execute_cmd(cmd)
> +            except Exception:
> +                logger.error('Failed to unpack NARR Grib data')
> +                raise
> +            finally:
> +                if len(output) > 0:
> +                    logger.info(output)
> +
> +
> +def extract_narr_aux_data(espa_metadata, aux_path):
> +    """Extracts the required NARR data from the auxiliary archive
> +
> +    Args:
> +        TODO TODO TODO
>
> same
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_extract_auxiliary_narr_data.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    for aux_set in aux_filenames(aux_path, PARMS_TO_EXTRACT,
> +                                 t0_date, t1_date):
> +
> +        logger.info('Using {0}'.format(aux_set.hdr))
> +        logger.info('Using {0}'.format(aux_set.grb))
> +
> +        # Verify that the files we need exist
> +        if (not os.path.exists(aux_set.hdr) or
> +                not os.path.exists(aux_set.grb)):
> +            raise Exception('Required LST AUX files are missing')
> +
> +        extract_from_grib(aux_set)
> +
> +
> +def main():
> +    """Core processing for ther application
>
> typo "ther", should be "the"
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_generate_products.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    parser.add_argument('--debug',
> +                        action='store_true', dest='debug',
> +                        required=False, default=False,
> +                        help='Output debug messages and/or keep debug data')
> +
> +    parser.add_argument('--version',
> +                        action='store_true', dest='version',
> +                        required=False, default=False,
> +                        help='Reports the version of the software')
> +
> +    args = parser.parse_args()
> +
> +    # Report the version and exit
> +    if args.version:
> +        print util.Version.version_text()
>
> same version comment
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_generate_products.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    # configuration
> +    if 'HOME' not in os.environ:
> +        raise Exception('[HOME] not found in environment')
> +    home_dir = os.environ.get('HOME')
> +
> +    # Build the full path to the configuration file
> +    config_path = os.path.join(home_dir, '.usgs', 'espa', filename)
> +
> +    return config_path
> +
> +
> +def retrieve_cfg(cfg_filename):
> +    """Retrieve the configuration for the cron
> +
> +    Returns:
> +        cfg <ConfigParser>: Configuration for ESPA cron
>
> Doesn't match parameter name.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_generate_products.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +        output = util.System.execute_cmd(' '.join(cmd))
> +    finally:
> +        if len(output) > 0:
> +            logger = logging.getLogger(__name__)
> +            logger.info(output)
> +
> +
> +def cleanup_intermediate_data():
> +    """Cleanup/remove all the LST temporary data
> +    """
> +
> +    # File cleanup
> +    cleanup_list = [GRID_POINT_HEADER_NAME,
> +                    GRID_POINT_BINARY_NAME
> +                   ]
>
> combine with the previous line
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_generate_products.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    cmd = ' '.join(cmd)
> +    output = ''
> +    try:
> +        logger.info('Calling [{0}]'.format(cmd))
> +        output = util.System.execute_cmd(cmd)
> +    except Exception:
> +        logger.error('Failed creating atmospheric parameters and generating '
> +                     'intermediate data')
> +        raise
> +    finally:
> +        if len(output) > 0:
> +            logger.info(output)
> +
> +    # Generate Land Surface Temperature band
> +    try:
> +        current_processor = build_lst_data.BuildLSTData( \
>
> backslash is not needed.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_generate_products.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +                     'intermediate data')
> +        raise
> +    finally:
> +        if len(output) > 0:
> +            logger.info(output)
> +
> +    # Generate Land Surface Temperature band
> +    try:
> +        current_processor = build_lst_data.BuildLSTData( \
> +            xml_filename=args.xml_filename)
> +        current_processor.generate_data()
> +    except Exception:
> +        logger.error('Failed processing Land Surface Temperature')
> +        raise
> +
> +
>
> remove one of these blank lines
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_run_modtran.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +    parser.add_argument('--debug',
> +                        action='store_true', dest='debug',
> +                        required=False, default=False,
> +                        help='Output debug messages and/or keep debug data')
> +
> +    parser.add_argument('--version',
> +                        action='store_true', dest='version',
> +                        required=False, default=False,
> +                        help='Reports the version of the software')
> +
> +    args = parser.parse_args()
> +
> +    # Report the version and exit
> +    if args.version:
> +        print util.Version.version_text()
>
> same version comment
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_run_modtran.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +        raise Exception('The MODTRAN data directory provided was empty')
> +
> +    return args
> +
> +
> +class Tape6rocessingError(Exception):
> +    """Exception specifically for Tape6 errors"""
> +    pass
> +
> +
> +TAPE6 = 'tape6'
> +TAPE6_SEARCH_STR = 'AREA-AVERAGED GROUND TEMPERATURE [K]'
> +
> +
> +def extract_target_surface_temp():
> +    """Swim in the pool using the path
>
> Please change this description. Think it was a copy paste from some test
> code.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_run_modtran.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +    # Report the version and exit
> +    if args.version:
> +        print util.Version.version_text()
> +        sys.exit(0)  # EXIT SUCCESS
> +
> +    if args.modtran_data_path is None:
> +        raise Exception('--modtran_data_path must be specified on the'
> +                        'command line')
> +
> +    if args.modtran_data_path == '':
> +        raise Exception('The MODTRAN data directory provided was empty')
> +
> +    return args
> +
> +
> +class Tape6rocessingError(Exception):
>
> Typo in exception name
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_run_modtran.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +        AREA-AVERAGED is easier to extract
> +    """
> +
> +    value = None
> +
> +    with open(TAPE6, 'r') as tape6_fd:
> +        for line in tape6_fd:
> +            # Remove all whitespace and newlines
> +            line = ' '.join(line.strip().split())
> +
> +            if line.startswith(TAPE6_SEARCH_STR):
> +                value = float(list(reversed(line.split()))[0])
> +                break
> +
> +    if value is None:
> +        raise Tape6rocessingError('Unable to find [{}] in tape6 file'
>
> typo in exception name
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_run_modtran.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +        radiance_hdr_fd.write('TARGET_PIXEL_SURFACE_TEMPERATURE {0}\n'
> +                              .format(tsp_value))
> +        radiance_hdr_fd.write('RADIANCE_RECORD_COUNT {0}\n'
> +                              .format(len(pltout_results)))
> +
> +
> +class ModtranProcessingError(Exception):
> +    """Exception specifically for MODTRAN errors"""
> +    pass
> +
> +
> +TAPE5 = 'tape5'
> +
> +
> +def process_point_dir((point_path, modtran_data_path)):
> +    """Swim in the pool using the path
>
> Again, this same comment was copied :(
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_utilities.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> >
>
> -import metadata_api
> +#import metadata_api
>
> remove this commented import
> ------------------------------
>
> In not-validated-prototype_lst/src/calculate_atmospheric_parameters.c
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +RETURN: double - The distance.
> +
> +NOTE: SR(x) = (scale_factor / cos ((x - false_easting) / equatorial_radius))
> +
> +NOTE: Simpson's Rule is applied for integrating the longitudinal distance
> +      from easting of first point to easting of second point.
> +
> +      SR(x)dx ~= ((e2 - e0) / 6)
> +                 * (SR(e0) + 4 * SR((e0 + e2) / 2) + SR(e2))
> +
> +      Where:
> +          e0 = easting of starting point
> +          e2 = easting of stopping point
> +******************************************************************************/
> +double distance_in_utm
>
> I think we should keep this distance implementation somewhere as an
> example of how to do it in UTM, But we should probably be using the
> haversine_distance, to be consistent with the python code, and to also
> allow us to use other projections. Maybe mixing Haversine in the python and
> UTM in the C is causing some of the differences.
> ------------------------------
>
> In not-validated-prototype_lst/src/calculate_atmospheric_parameters.c
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +        for (point = 0; point < NUM_CELL_POINTS; point++)
> +        {
> +            parameters[parameter] += (w[point] * at_height[point][parameter]);
> +        }
> +    }
> +}
> +
> +
> +/*****************************************************************************
> +METHOD:  determine_grid_point_distances
> +
> +PURPOSE: Determines the distances for the current set of grid points.
> +
> +NOTE: The indexes of the grid points are assumed to be populated.
> +*****************************************************************************/
> +void determine_grid_point_distances
>
> This code changes to use the haversine distance.
> ------------------------------
>
> In not-validated-prototype_lst/src/calculate_atmospheric_parameters.c
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +
> +/*****************************************************************************
> +METHOD:  determine_center_grid_point
> +
> +PURPOSE: Determines the index of the center point from the current set of grid
> +         points.
> +
> +NOTE: The indexes of the grid points are assumed to be populated.
> +
> +RETURN: type = int
> +    Value  Description
> +    -----  -------------------------------------------------------------------
> +    index  The index of the center point
> +*****************************************************************************/
> +int determine_center_grid_point
>
> maybe haversine modifications filter through here as well
> ------------------------------
>
> In not-validated-prototype_lst/src/calculate_atmospheric_parameters.c
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +
> +/*****************************************************************************
> +METHOD:  determine_first_center_grid_point
> +
> +PURPOSE: Determines the index of the first center point to use for the current
> +         line.  Only called when the fist valid point for a line is
> +         encountered.  The point is determined from all of the available
> +         points.
> +
> +RETURN: type = int
> +    Value  Description
> +    -----  -------------------------------------------------------------------
> +    index  The index of the center point
> +*****************************************************************************/
> +int determine_first_center_grid_point
>
> maybe haversine modifications filter through here as well
> ------------------------------
>
> In not-validated-prototype_lst/src/calculate_atmospheric_parameters.c
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>
> :
>
> > +
> +        pixel_line_loc = line * input->samples;
> +
> +        /* Set first_sample to be true */
> +        first_sample = true;
> +        for (sample = 0; sample < input->samples; sample++)
> +        {
> +            pixel_loc = pixel_line_loc + sample;
> +
> +            if (inter.band_thermal[pixel_loc] != LST_NO_DATA_VALUE)
> +            {
> +                /* Determine UTM coordinates for current line/sample */
> +                easting = input->meta.ul_map_corner.x
> +                    + (sample * input->x_pixel_size);
> +                northing = input->meta.ul_map_corner.y
> +                    - (line * input->y_pixel_size);
>
> Haversine changes start here probably, need to calculate lat and lon to
> pass through to haversine code.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-12686423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR0JYSsC29A7AhyEf4acyDr5DqXf3D4-ks5rHtgKgaJpZM4LK7sG>
> .
>



-- 

Ray Dittmeier
Scientific Software Engineer
Stinger Ghaffarian Technologies (SGT, Inc.)
Contractor to the U.S. Geological Survey (USGS)
Earth Resources Observation and Science (EROS) Center
47914 252nd Street
Sioux Falls, SD 57198
Phone: 605-594-6898
ray.dittmeier.ctr@usgs.gov <firstname.lastname.ctr@usgs.gov>

gschmidt-usgs commented on README.md:
```diff
@@ -1,27 +1,31 @@
 
-## Land Surface Temperature 0.0.4 Release Notes
-Release Date: August 2016
+## Land Surface Temperature 0.1.1 Release Notes
```
Should this be 0.1.0 versus 0.1.1?

gschmidt-usgs commented on README.md:
```diff
@@ -71,15 +75,15 @@ See `land_surface_temperature.py --help` for command line details.
-The following input data are required to generate the spectral indicies products:
+The following input data are required to generate the Land Surface Temperature product:
 * Top of Atmosphere Reflectance (TOA)
   - TOA products can be generated using the [LEDAPS](https://github.com/USGS-EROS/espa-surface-reflectance) or [L8_SR](https://github.com/USGS-EROS/espa-surface-reflectance) software found in our [espa-surface-reflectance](https://github.com/USGS-EROS/espa-surface-reflectance) project.  Or through our ondemand processing system [ESPA](https://espa.cr.usgs.gov), be sure to select the ENVI output f ...
```
The "Or through our ondemand processing system ..." is not a sentence by itself and really belongs to the previous sentence.  Then make the "be sure to select the ENVI output format" its own sentence.

raydittmeier replied:
I changed this to use full sentences.

gschmidt-usgs commented on README.md:
```diff
@@ -71,15 +75,15 @@ See `land_surface_temperature.py --help` for command line details.
   - Elevation data is expected to be in the same projection and image dimensions as the TOA products and in the ENVI file format.  The filename is also expected be `<sceneid>_dem.img`: where <b>sceneid</b> is the Landsat Scene ID of the data being processed.
 * ASTER GED
-  - ASTER GED data can be [found here](https://lpdaac.usgs.gov/data_access/data_pool).  However it will automatically be retrieved for you as needed and cleaned up.
+  - ASTER GED data can be [found here](https://lpdaac.usgs.gov/data_access/data_pool).  External users will need to set up an alternate ASTER GED data retrieval method.  One potential method is to create a NASA Earthdata account has [described here] (http://e4ftl01.cr.usgs.gov/ASTT), and update the automated LST procedure to supply the login information before accessing the ASTER GED data.
```
"create a NASA Earthdata account has [described here]"   The word 'has' doesn't fit.  Should this be 'as'?  Otherwise this is a good rewrite of the explanation for the ASTER GED!

raydittmeier replied:
It was a typo.  I changed it to "as".

gschmidt-usgs commented on narr_tools/extract_archived_3d.py:
```diff
@@ -0,0 +1,218 @@
+    return parser.parse_args()
+
+
+def process_parameter(filename, output_path, parameter, hour):
```
Many of these methods are missing the standard function headers to describe the function and the input/output parameters.

ronaldddilley replied:
Since this is analysis code, intended to be created as quick and dirty as possible, I didn't take the effort to clean it up.

gschmidt-usgs commented on narr_tools/extract_archived_3d.py:
```diff
@@ -0,0 +1,218 @@
+
+
+def process_file(filename):
+    print filename
```
This is probably left-over debugging info?

ronaldddilley replied:
Since this is analysis code, intended to be created as quick and dirty as possible, I didn't take the effort to clean it up.

gschmidt-usgs commented on not-validated-prototype_lst/README.md:
```diff
@@ -1,3 +1,2 @@
-
-However most of the information can be found in the top-level README.
+This algorithm-specific README will be used if multiple LST algorithms are 
+implemented.  Currently the top-level README has the required information.
```
Should we just delete this README?

raydittmeier replied:
Originally we were thinking it would be better to leave in a stub file, but now that doesn't seem like a good idea to me, so I removed it as suggested.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/determine_missing_narr.py:
```diff
@@ -0,0 +1,289 @@
+
+# We can use strings for these in this code to make things simpler for
+# directory creation
+# The temperatures we run each elevation through
```
Are these Kelvin?

raydittmeier replied:
Yes, LST generally uses Kelvin.  I updated the comment.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/determine_missing_narr.py:
```diff
@@ -0,0 +1,289 @@
+
+    Args:
+        parameter <str>: The parameter to load
+        layer <str>: The pressure layer to load for the parameter
```
I like how you define the data type for these arguments.  Might have to look at doing that with my code.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/determine_missing_narr.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def load_narr_pressure_layers(parameters, layers):
+    """Loads all the parameters and presssure layes
```
pressssure ==> pressure
layes ==> layers

raydittmeier replied:
I fixed the typos.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/determine_missing_narr.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def check_hgt(data, point, layer, time):
+    """Determines"""
```
Determines .....??

raydittmeier replied:
I filled in the function description here and in some other nearby spots.  I also added comments describing the function parameters.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/determine_missing_narr.py:
```diff
@@ -0,0 +1,289 @@
+    # Load the grid information
+    (grid_points, dummy1, dummy2) = read_grid_points()
+
+    determine_missing_data(grid_points=grid_points)
```
Does anything happen with the missing points information besides them getting written to stdout via the print statements? Or does this get dumped to a file that is then used by MODTRAN or something?

raydittmeier replied:
I think this is just a diagnostic tool, not used by the main LST processing code.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py:
```diff
@@ -29,1007 +29,1245 @@
+            bi_swir1 = get_band_info(band)
+
+        if (band.get('product') == 'toa_bt' and
+                band.get('category') == 'image'):
```
This works now, but could break if another band were added to the BT product.  Why not look for band.get('name' == 'toa_band6' versus the band.get('category') == 'image'?

raydittmeier replied:
I changed this to use band.get('name') with cases for L8 and the other satellites.  I ran an L8 test and a couple non-L8 tests to make sure the results didn't change.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py:
```diff
@@ -29,1007 +29,1245 @@
+    red_data = red_data * src_info.toa.red.scale_factor
+
+    # NDVI ---------------------------------------------------------------
+    ndvi_data = ((nir_data - red_data) / (nir_data + red_data))
```
Given that both the NIR and red bands have the same scale factor, they technically don't have to be unscaled to compute the NDVI.  5000/10000 = 0.5 --> 5/10 = 0.5
If you chose to skip the unscaling, then the same would apply in the NDSI computations below.  Then the comparison with 0.4 to determine snow locations would need to be 0.4 * the scale factor.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py:
```diff
@@ -29,1007 +29,1245 @@
+    Args:
+        coefficients <CoefficientInfo>: coefficients for the math
+        tile_name <str>: Filename to create for the tile
+        aster_b13_data <numpy.2darray>: Band 13 ASTER tile data
```
It might be good to specify (unscaled) in the comment so the user knows the data is not exactly as it was read from the file.

raydittmeier replied:
I added unscaled to the comment.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/lst_determine_grid_points.py:
```diff
@@ -0,0 +1,772 @@
+    """
+
+    # Adjustment is specific to NARR data resolution in meters
+    adjustment = 32000 * 1.0
```
why multiply by 1.0?

gschmidt-usgs commented on not-validated-prototype_lst/scripts/lst_determine_grid_points.py:
```diff
@@ -0,0 +1,772 @@
+        lat <float>: The latitude to check
+
+    Returns:
+        <boo>:  True - If within the bounds
```
<bool>

raydittmeier replied:
I fixed the "boo" comment in a couple spots.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/lst_generate_products.py:
```diff
@@ -0,0 +1,368 @@
+
+    output = ''
+    try:
+        cmd = ['lst_determine_grid_points.py',
```
Can't we call the main methods in these Python apps as a function call versus kicking off a Python executable?

raydittmeier replied:
I don't think it will work that way without changing the command line tools to be more like libraries.  I tried importing one of the tools, and then calling main, but main didn't accept an input parameter, and we need input parameters when calling these tools.

gschmidt-usgs commented on not-validated-prototype_lst/scripts/lst_utilities.py:
```diff
@@ -29,7 +27,7 @@ class Version(object):
     '''
 
-    version = '0.0.4'
+    version = '0.1.1'
```
If the version in the README is changed to 0.1.0 than this should change too.  Not sure why the jump to 0.1.1 from 0.0.4.

gschmidt-usgs commented:
No further comments.

ronaldddilley commented:
I did that just to show using multiples of the NARR resolution.

On Mon, Dec 19, 2016 at 4:49 PM, gschmidt-usgs <notifications@github.com>
wrote:

> *@gschmidt-usgs* commented on this pull request.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_determine_grid_points.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13659353>
> :
>
> > +                    nlines=nlines, nsamps=nsamps, fill_value=fill_value)
> +
> +
> +def determine_adjusted_data_bounds(espa_metadata, gdal_objs):
> +    """Determines the data boundaries specific to the NARR data resolution
> +
> +    Args:
> +        espa_metadata <espa.Metadata>: The metadata for the data
> +        gdal_objs <GdalInfo>: Contains GDAL objects and static information
> +
> +    Returns:
> +        <DataBoundInfo>: Contains adjusted data boundry information
> +    """
> +
> +    # Adjustment is specific to NARR data resolution in meters
> +    adjustment = 32000 * 1.0
>
> why multiply by 1.0?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13659353>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALk-cpTA3_jO7KZJINp6qknmgwK66vv0ks5rJwoPgaJpZM4LK7sG>
> .
>



-- 
---------------------------------------------------------------
Ronald D. Dilley Jr.

Innovate!, Inc.

Contractor To:
  USGS Earth Resources Observation and Science Center (EROS)
  47914 252nd Street
  Sioux Falls, SD 57198-9801

Phone: 605.594.2669
Email: ronald.dilley.ctr@usgs.gov
Email: rdilley@innovateteam.com
http://landsat.usgs.gov/CDR_ECV.php
---------------------------------------------------------------

raydittmeier commented:
Per Ron, we want to match the version in Jira, and 0.1.0 was already used
for something else there (something that was already closed).  That's why
0.1.1 was selected.

On Mon, Dec 19, 2016 at 2:35 PM, gschmidt-usgs <notifications@github.com>
wrote:

> *@gschmidt-usgs* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13636101>
> :
>
> > @@ -1,27 +1,31 @@
>  ## NOTE - The products produced by this software have not been validated and are considered prototype.
>
> -## Land Surface Temperature 0.0.4 Release Notes
> -Release Date: August 2016
> +## Land Surface Temperature 0.1.1 Release Notes
>
> Should this be 0.1.0 versus 0.1.1?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13636101>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR0JYc37rsslKFw74GR6APGUymo474lIks5rJuqigaJpZM4LK7sG>
> .
>



-- 

Ray Dittmeier
Scientific Software Engineer
Stinger Ghaffarian Technologies (SGT, Inc.)
Contractor to the U.S. Geological Survey (USGS)
Earth Resources Observation and Science (EROS) Center
47914 252nd Street
Sioux Falls, SD 57198
Phone: 605-594-6898
ray.dittmeier.ctr@usgs.gov <firstname.lastname.ctr@usgs.gov>

raydittmeier commented:
Per Ron, we want to match the version in Jira, and 0.1.0 was already used
for something else there (something that was already closed).  That's why
0.1.1 was selected.

On Mon, Dec 19, 2016 at 4:59 PM, gschmidt-usgs <notifications@github.com>
wrote:

> *@gschmidt-usgs* commented on this pull request.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/lst_utilities.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13660867>
> :
>
> > @@ -29,7 +27,7 @@ class Version(object):
>          Provides methods for retrieving version information.
>      '''
>
> -    version = '0.0.4'
> +    version = '0.1.1'
>
> If the version in the README is changed to 0.1.0 than this should change
> too. Not sure why the jump to 0.1.1 from 0.0.4.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13660867>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR0JYQ9NLeFjuav0mLR9YD4JIrBgxadgks5rJwxigaJpZM4LK7sG>
> .
>



-- 

Ray Dittmeier
Scientific Software Engineer
Stinger Ghaffarian Technologies (SGT, Inc.)
Contractor to the U.S. Geological Survey (USGS)
Earth Resources Observation and Science (EROS) Center
47914 252nd Street
Sioux Falls, SD 57198
Phone: 605-594-6898
ray.dittmeier.ctr@usgs.gov <firstname.lastname.ctr@usgs.gov>

ronaldddilley commented on narr_tools/extract_merged_AWIP32_3d.py:
```diff
@@ -0,0 +1,221 @@
+                        help='Debug logging')
+
+    return parser.parse_args()
+
```
They are actually taken from the main code base and copied here.  But since this is analysis code, intended to be created as quick and dirty as possible, I didn't take the effort to clean it up.

raydittmeier commented:
I'm reluctant to do this one because it wouldn't work if the scaling factor
is different for the 2 NDVI (or NDSI) bands.  Ron says the scaling factor
should be the same for the reflective bands, but I don't know if that would
always hold, or if someone would change the metadata values in a test.
Also, right now nir_data and red_data aren't used for anything else, but
maybe the code could change in that respect in the future, in which case
the scaling factor should be applied to them.  Also, I'd be concerned that
someone would copy the code and use invalid nir_data/red_data values.  If
this code were slow I'd lean more towards optimizing it, but the MODTRAN
phase is the one that takes a long time.

On Mon, Dec 19, 2016 at 4:14 PM, gschmidt-usgs <notifications@github.com>
wrote:

> *@gschmidt-usgs* commented on this pull request.
> ------------------------------
>
> In not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13653717>
> :
>
> >
> +    logger.info('Building TOA based NDVI band for Landsat data')
> +
> +    # NIR ----------------------------------------------------------------
> +    nir_data = extract_raster_data(src_info.toa.nir.name, 1)
> +    nir_no_data_locations = np.where(nir_data == no_data_value)
> +    nir_data = nir_data * src_info.toa.nir.scale_factor
> +
> +    # RED ----------------------------------------------------------------
> +    red_data = extract_raster_data(src_info.toa.red.name, 1)
> +    red_no_data_locations = np.where(red_data == no_data_value)
> +    red_data = red_data * src_info.toa.red.scale_factor
> +
> +    # NDVI ---------------------------------------------------------------
> +    ndvi_data = ((nir_data - red_data) / (nir_data + red_data))
>
> Given that both the NIR and red bands have the same scale factor, they
> technically don't have to be unscaled to compute the NDVI. 5000/10000 = 0.5
> --> 5/10 = 0.5
> If you chose to skip the unscaling, then the same would apply in the NDSI
> computations below. Then the comparison with 0.4 to determine snow
> locations would need to be 0.4 * the scale factor.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-EROS/espa-land-surface-temperature/pull/11#pullrequestreview-13653717>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR0JYaUASbUnuamYYT9B83WV1zvpYnSTks5rJwHAgaJpZM4LK7sG>
> .
>



-- 

Ray Dittmeier
Scientific Software Engineer
Stinger Ghaffarian Technologies (SGT, Inc.)
Contractor to the U.S. Geological Survey (USGS)
Earth Resources Observation and Science (EROS) Center
47914 252nd Street
Sioux Falls, SD 57198
Phone: 605-594-6898
ray.dittmeier.ctr@usgs.gov <firstname.lastname.ctr@usgs.gov>

raydittmeier commented on not-validated-prototype_lst/scripts/estimate_landsat_emissivity.py:
```diff
@@ -29,1007 +29,1245 @@
+
+    Args:
+        tile_name <str>: Filename to create for the tile
+        ndvi_data <numpy.2darray>: NDVI Band ASTER tile data
```
I added unscaled to the comment.
